### DISCUSSION TITLE
Documentation enhancement (readme.md)

### DISCUSSION THREAD

**Discussion by Verhoeckx** [nextcloud/all-in-one#1744] [Questions] [answered]
### Is your feature request related to a problem? Please describe. 
I just started reading the documentation and missed the following:

1. The documentation doesn't describe what `--sig-proxy=false` does (in the section `Explanation of the command`).
2. I miss a diagram of all the containers that are spawned and how they are connected to each other. This could give me (and others) valuable insight.
3. The section `Which ports are mandatory to be open in your firewall/router?` describes three port (443, 3478/tcp and 3478/udp), but then starts describing _six_ port (8080, 80, 8443, 443, 3478/tcp and 347/udp. Now which are truly mandatory? That's unclear to me.

### Describe the solution you'd like
It would be great if the documentation would answer question 1 and 3. A diagram of all spawned containers, and how they are connected with each other, would also be very helpful.

**Comment by ManOki** [comment, marked as answer]
>     1. The documentation doesn't describe what `--sig-proxy=false` does (in the section `Explanation of the command`).

This is a `docker run` [command option](https://docs.docker.com/engine/reference/commandline/container_run/); fixed with https://github.com/nextcloud/all-in-one/pull/1742

>     2. I miss a diagram of all the containers that are spawned and how they are connected to each other. This could give me (and others) valuable insight.

See [manual-install](https://github.com/nextcloud/all-in-one/tree/main/manual-install), especially the latest.yml (aka. docker-compose.yml). All containers except the mastercontainer are listed there including their ports. Note that not every container is mandatory, e.g., you don't need imaginary. So there are a lot of variants. The internal connection may be revealed from the `depends_on` part of the latest.yml file.

>     3. The section `Which ports are mandatory to be open in your firewall/router?` describes three port (443, 3478/tcp and 3478/udp), but then starts describing _six_ port (8080, 80, 8443, 443, 3478/tcp and 347/udp. Now which are truly mandatory? That's unclear to me.

The three ports (443, 3478 tcp+udp) have to be available outside of your host machine/home network to make nextcloud itself work correctly. So you _have to_ open them in your firewall or router. The other ports are optional due to their described functionality:

- if you have no other setup, you might directly use the internal ACME functionality of AIO to get a valid SSL certificate and therefore need to open port 80
- if you want access to the mastercontainer outside of your host machine you have to open port 8443
- other port 8080 is 'just' for internal reason/initial setup and should not be open to the public (use 8443 for public access)

**Comment by Verhoeckx** [comment]
Hello @ManOki,

Thank you very much for the elaborate answer! There are still some  parts I do not understand, but I will look them up. Using the `lastest.yml` file I will create a simple diagram of all the containers that are getting spawned. I will send the svg-file and then you can decide to use it or not.

And oh, I think it's a good idea to move the section `How to change the default location of Nextcloud's Datadir?` more to the top of the readme file, because `NEXTCLOUD_DATADIR` must be added to the initial startup (as described) of Nextcloud AIO.

**Reply by szaimen** [reply]
Does https://github.com/nextcloud/all-in-one/pull/1756 look good to you?

**Comment by Verhoeckx** [comment]
This is how I imagine Nextcloud AIO works:

![Nextcloud AIO - containers creation](https://user-images.githubusercontent.com/64807887/212419683-e6442e7a-8cdd-455a-b289-e1fa76a7e0f7.png)


Let me know is you want to have the SVG-file (I can't upload it here because then the file gets corrupted).

**Reply by Zoey2936** [reply]
I think that's not really correct

**Reply by Verhoeckx** [reply]
You mean the sequence of the creation of the containers?

**Reply by Zoey2936** [reply]
The mastercontainer creates this container, but the mastercontainer only managed them, they arent conected to the masterconatiner

**Reply by Verhoeckx** [reply]
With the arrows I mean that they are created ('spawned') by the Mastercontainer. Not how they are connected internally. This should be clearer?

**Comment by Verhoeckx** [comment]
Instead of showing a diagram of all the containers that are created, I think it's enough to list the containers that are created in the section `How does it work?`.

**For example:**

This concept allows a user to install only one container with a single command that does the heavy lifting of creating and managing all `the` containers that are needed in order to provide a Nextcloud installation <strike>with most features included</strike>

When you run the Docker command the following containers are created:
- nextcloud-aio-mastercontainer (the container that creates all the other containers)
- nextcloud-aio-apache
- nextcloud-aio-database (uses PostgreSQL)
- nextcloud-aio-nextcloud
- nextcloud-aio-redis (used for caching Nextcloud in the RAM memory)
- nextcloud-aio-collabora (the back end of Nextcloud Office) 
- nextcloud-aio-talk
- nextcloud-aio-clamav (open source antivirus software)
- nextcloud-aio-onlyoffice
- nextcloud-aio-imaginary (used for creating image previews and thumbnails)
- nextcloud-aio-fulltextsearch (used for doing a full text search in Nextcloud)

**Comment by salleq** [comment]
I banged my head yesterday trying to get AIO working with an existing Traefik. I did finally manage to do that with the manual install, but I'd want to avoid going the manual update route in the future.

My setup is in a way that I cannot get through port 443 from outside, so I could only use that internally (I have 8443 open from outside though)

There's some more stuff that might help people like me in the readme.md, like below:
"...because port 443 is used by this project and opened on the host by default even though it does not look like this is the case".

There's no explanation WHY port 443 is used, but here's something I've gathered, could be wrong:

- Port 443 from INTERNET  to the wanted DNS name, say nextcloud.wherever.com:443 needs to be routed to the AIO containers Apache port (let's say TCP/11000).
- Port 443 is hardcoded, so I can't install AIO with nextcloud.wherever.com:8443 (even though that port would go to aforementioned Apache port and I get the response from there that AIO installation is expecting)
- You can't use LAN IP addresses (yeah the error " It seems like the ip-address is set to an internal or reserved ip-address. This is not supported." kind of clearly states it, but I didn't get it at first, it's because I have my internal DNS set to a LAN IP...

So, I do get the requirement of not using LAN IPs (kind of), but what about changing the port where the AIO container tries to contact? Or is this something that is going to be supported either?

**Reply by salleq** [reply]
Funny thing, I got the LAN part working with a FQDN (even though it's pointing to a LAN IP) after all. So my traffic in LAN goes from client -> traefik -> AIO (both on the same docker host) with https and port 443, but while I tried to get it working like I previously had it, through the internet with the same FQDN but with different port (8443), I guess AIO forces redirect to port 443 so the client outside my firewall gets redirected to a port that is not open - and this redirection is something that cannot be changed? 

EDIT: re-re-read documentation and I gathered port 443 is required to the container, so I'm out of luck here!
